File: AccuPay/Domain/Paystubs/Product.vb
﻿Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema

Namespace Global.AccuPay.Entity

    <Table("product")>
    Public Class Product

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Overridable Property RowID As Integer?

        Public Overridable Property SupplierID As Integer?

        Public Overridable Property Name As String

        Public Overridable Property OrganizationID As Integer?

        Public Overridable Property Description As String

        Public Overridable Property PartNo As String

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Overridable Property Created As Date

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Overridable Property LastUpd As Date?

        <NotMapped>
        Public Overridable Property LastArrivedQty As Integer?

        Public Overridable Property CreatedBy As Integer?

        Public Overridable Property LastUpdBy As Integer?

        Public Overridable Property Category As String

        <ForeignKey("CategoryID")>
        Public Overridable Property CategoryEntity As Category

        Public Overridable Property CategoryID As Integer?

        <NotMapped>
        Public Overridable Property AccountingAccountID As Integer?

        <NotMapped>
        Public Overridable Property Catalog As String

        <NotMapped>
        Public Overridable Property Comments As String

        Public Overridable Property Status As String

        Public Overridable Property Fixed As Boolean

        <NotMapped>
        Public Overridable Property UnitPrice As Decimal

        <NotMapped>
        Public Overridable Property VATPercent As Decimal

        <NotMapped>
        Public Overridable Property FirstBillFlag As Char

        <NotMapped>
        Public Overridable Property SecondBillFlag As Char

        <NotMapped>
        Public Overridable Property ThirdBillFlag As Char

        <NotMapped>
        Public Overridable Property PDCFlag As Char

        <NotMapped>
        Public Overridable Property MonthlyBIllFlag As Char

        <NotMapped>
        Public Overridable Property PenaltyFlag As Char

        <NotMapped>
        Public Overridable Property WithholdingTaxPercent As Decimal

        <NotMapped>
        Public Overridable Property CostPrice As Decimal

        <NotMapped>
        Public Overridable Property UnitOfMeasure As String

        <NotMapped>
        Public Overridable Property SKU As String

        <NotMapped>
        Public Overridable Property LeadTime As Integer?

        <NotMapped>
        Public Overridable Property BarCode As String

        <NotMapped>
        Public Overridable Property BusinessUnitID As Integer?

        <NotMapped>
        Public Overridable Property LastRcvdFromShipmentDate As Date

        <NotMapped>
        Public Overridable Property LastRcvdFromShipmentCount As Integer?

        <NotMapped>
        Public Overridable Property TotalShipmentCount As Integer?

        <NotMapped>
        Public Overridable Property BookPageNo As String

        <NotMapped>
        Public Overridable Property BrandName As String

        <NotMapped>
        Public Overridable Property LastPurchaseDate As Date

        <NotMapped>
        Public Overridable Property LastSoldDate As Date

        <NotMapped>
        Public Overridable Property LastSoldCount As Integer?

        <NotMapped>
        Public Overridable Property ReOrderPoint As Integer?

        Public Overridable Property AllocateBelowSafetyFlag As Char

        <NotMapped>
        Public Overridable Property Strength As String

        <NotMapped>
        Public Overridable Property UnitsBackordered As Integer?

        <NotMapped>
        Public Overridable Property UnitsBackorderAsOf As Date

        <NotMapped>
        Public Overridable Property DateLastInventoryCount As Date

        <NotMapped>
        Public Overridable Property TaxVAT As Decimal

        <NotMapped>
        Public Overridable Property WithholdingTax As Decimal

        <NotMapped>
        Public Overridable Property COAId As Integer?

        '<NotMapped>
        Public Overridable Property ActiveData As Boolean

        Public Overridable ReadOnly Property IsTaxable As Boolean
            Get
                Return Status = "1"
            End Get
        End Property

        Public ReadOnly Property IsVacationOrSickLeave As Boolean
            Get
                Return IsVacationLeave OrElse IsSickLeave
            End Get
        End Property

        Public ReadOnly Property IsVacationLeave As Boolean
            Get
                Return PartNo.Trim.ToUpper = ProductConstant.VACATION_LEAVE_PART_NO.ToUpper
            End Get
        End Property

        Public ReadOnly Property IsSickLeave As Boolean
            Get
                Return PartNo.Trim.ToUpper = ProductConstant.SICK_LEAVE_PART_NO.ToUpper
            End Get
        End Property

        Public ReadOnly Property IsThirteenthMonthPay As Boolean
            Get
                Return AllocateBelowSafetyFlag = "1"
            End Get
        End Property

    End Class

End Namespace

File: Accupay.Data/Entities/Product.cs
using AccuPay.Data.Helpers;
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("product")]
    public class Product
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int? RowID { get; set; }

        public string Name { get; set; }

        public int? OrganizationID { get; set; }

        public string Description { get; set; }

        public string PartNo { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public DateTime Created { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        public DateTime? LastUpd { get; set; }

        public int? CreatedBy { get; set; }

        public int? LastUpdBy { get; set; }

        public string Category { get; set; }

        [ForeignKey("CategoryID")]
        public Category CategoryEntity { get; set; }

        public int? CategoryID { get; set; }

        public string Comments { get; set; }

        public string Status { get; set; }

        public bool Fixed { get; set; }

        public char AllocateBelowSafetyFlag { get; set; }

        public bool ActiveData { get; set; }

        public string DisplayName => PartNo;

        public bool IsTaxable => Status == "1";

        public bool IsVacationOrSickLeave => IsVacationLeave || IsSickLeave;

        public bool IsVacationLeave => PartNo.Trim().ToUpper() == ProductConstant.VACATION_LEAVE.ToUpper();

        public bool IsSickLeave => PartNo.Trim().ToUpper() == ProductConstant.SICK_LEAVE.ToUpper();

        public bool IsMaternityLeave => PartNo.Trim().ToUpper() == ProductConstant.MATERNITY_LEAVE.ToUpper();

        public bool IsParentalLeave => PartNo.Trim().ToUpper() == ProductConstant.PARENTAL_LEAVE.ToUpper();

        public bool IsOthersLeave => PartNo.Trim().ToUpper() == ProductConstant.OTHERS_LEAVE.ToUpper();

        public bool IsPagibigLoan => PartNo.Trim().ToUpper() == ProductConstant.PAG_IBIG_LOAN.ToUpper();

        public bool IsSssLoan => PartNo.Trim().ToUpper() == ProductConstant.SSS_LOAN.ToUpper();

        public bool IsThirteenthMonthPay => AllocateBelowSafetyFlag == '1';
    }
}
